include/boost/windows_sspi/detail/sspi_impl.hpp

```diff
@@ -371,12 +378,93 @@ class sspi_decrypt {
   SECURITY_STATUS m_last_error;
 };
 
+class sspi_shutdown {
+public:
+  enum class state {
+    data_available,
+    error
+  };
+
+  sspi_shutdown(CtxtHandle* context, CredHandle* credentials)
+    : m_context(context)
+    , m_credentials(credentials)
+    , m_last_error(SEC_E_OK) {
+  }
+
+  state operator()() {
+    SecBufferDesc OutBuffer;
+    SecBuffer OutBuffers[1];
+
+    DWORD shutdown_type = SCHANNEL_SHUTDOWN;
+
+    OutBuffers[0].pvBuffer = &shutdown_type;
+    OutBuffers[0].BufferType = SECBUFFER_TOKEN;
+    OutBuffers[0].cbBuffer = sizeof(shutdown_type);
+
+    OutBuffer.cBuffers  = 1;
+    OutBuffer.pBuffers  = OutBuffers;
+    OutBuffer.ulVersion = SECBUFFER_VERSION;
+
+    m_last_error = detail::sspi_functions::ApplyControlToken(m_context, &OutBuffer);
+    if (m_last_error != SEC_E_OK) {
+      return state::error;
+    }
+
+    OutBuffers[0].pvBuffer   = NULL;
+    OutBuffers[0].BufferType = SECBUFFER_TOKEN;
+    OutBuffers[0].cbBuffer   = 0;
+
+    OutBuffer.cBuffers  = 1;
+    OutBuffer.pBuffers  = OutBuffers;
+    OutBuffer.ulVersion = SECBUFFER_VERSION;
+
+    DWORD out_flags = 0;
+
+    m_last_error = detail::sspi_functions::InitializeSecurityContext(m_credentials,
+                                                                     m_context,
+                                                                     NULL,
+                                                                     context_flags,
+                                                                     0,
+                                                                     SECURITY_NATIVE_DREP,
+                                                                     NULL,
+                                                                     0,
+                                                                     m_context,
+                                                                     &OutBuffer,
+                                                                     &out_flags,
+                                                                     nullptr);
+    if (m_last_error != SEC_E_OK) {
+      return state::error;
+    }
+
+    // TODO: Avoid this copy
+    m_output_data = std::vector<char>{reinterpret_cast<const char*>(OutBuffers[0].pvBuffer),
+      reinterpret_cast<const char*>(OutBuffers[0].pvBuffer) + OutBuffers[0].cbBuffer};
+    detail::sspi_functions::FreeContextBuffer(OutBuffers[0].pvBuffer);
+    return state::data_available;
+  }
+
+  boost::system::error_code last_error() const {
+    return error::make_error_code(m_last_error);
+  }
+
+  std::vector<char> data() const {
```

Comment:
The copy could and should be avoided, but as with the potential memory leak above, I prefer to keep it as it is for now.

It is getting pretty clear, I think, that the classes in `sspi_impl` are sharing some common functionality/traits/stuff and I hope to somehow refactor that so they can share some functionality and avoid duplication.

Not really sure how to handle that, but I hope it will become clearer once I have implemented the missing functionality.

Again, I hope that makes sense, please feel free to tell me if you think this might be the wrong approach.